src/platform_impl/linux/wayland/window/mod.rs

```diff
@@ -0,0 +1,643 @@
+        // Max dimensions.
+        let max_size = attributes
+            .min_inner_size
+            .map(|size| size.to_logical::<f64>(scale_factor as f64).into());
+        window.set_max_size(max_size);
+
+        // Set Wayland specific window attributes.
+        if let Some(app_id) = platform_attributes.app_id {
+            window.set_app_id(app_id);
+        }
+
+        // Set common window attributes.
+        //
+        // We set resizable after other attributes, since it touches min and max size under
+        // the hood.
+        window.set_resizable(attributes.resizable);
+        window.set_title(attributes.title);
+
+        // Set fullscreen/maximized if so was requested.
+        match attributes.fullscreen {
+            Some(Fullscreen::Exclusive(_)) => {
+                panic!("Wayland doesn't support exclusive fullscreen.")
+            }
+            Some(Fullscreen::Borderless(monitor)) => {
+                let monitor =
+                    monitor.and_then(|RootMonitorHandle { inner: monitor }| match monitor {
+                        PlatformMonitorHandle::Wayland(monitor) => Some(monitor.proxy),
+                        #[cfg(feature = "x11")]
+                        PlatformMonitorHandle::X(_) => None,
+                    });
+
+                window.set_fullscreen(monitor.as_ref());
+            }
+            None => {
+                if attributes.maximized {
+                    window.set_maximized();
+                }
+            }
+        }
+
+        let size = Arc::new(Mutex::new(LogicalSize::new(width, height)));
+
+        // We should trigger redraw and commit the surface for the newly created window.
+        let mut window_update = WindowUpdate::new();
+        window_update.refresh_frame = true;
+        window_update.redraw_requested = true;
+
+        let window_id = super::make_wid(&surface);
+        let window_requests = Arc::new(Mutex::new(Vec::with_capacity(64)));
+
+        // Create a handle that performs all the requests on underlying sctk a window.
+        let window_handle = WindowHandle::new(window, size.clone(), window_requests.clone());
+
+        event_loop_window_target
+            .window_map
+            .borrow_mut()
+            .insert(window_id, window_handle);
+
+        event_loop_window_target
+            .window_updates
+            .borrow_mut()
+            .insert(window_id, WindowUpdate::new());
+
+        let windowing_features = event_loop_window_target.windowing_features;
+
+        // Send all updates to the server.
+        let wayland_source = &event_loop_window_target.wayland_source;
+        let mut winit_state = event_loop_window_target.get_winit_state();
+        let event_loop_handle = &event_loop_window_target.event_loop_handle;
+
+        event_loop_handle.with_source(&wayland_source, |event_queue| {
+            let event_queue = event_queue.queue();
+            let _ = event_queue.sync_roundtrip(&mut winit_state, |_, _, _| unreachable!());
```

Comment:
What's this?

Reply:
it bootstraps a window, aka tells compositor that we created something so it'll issue a Configure for a window, since you can't really touch a window unless you acked a configure compositor sent to you.

Reply:
Does `sync_roundtrip` block then? When is the callback called, will `unreachable` truly never be called? What's the point of the closure then? Can sctk not do this?

Just sprinkling `sync_roundtrip` everywhere just seems a bit strange and hard to actually understand for people reading the code why it is needed in the specific place it is used.

Reply:
>Does sync_roundtrip block then? When is the callback called, will unreachable truly never be called? What's the point of the closure then? Can sctk not do this?

All objects which closures wasn't assigned with `quick_assign` or some other things will go there, this is not part of sctk, but is a part of wayland-client, hence reexport.

>Just sprinkling sync_roundtrip everywhere just seems a bit strange and hard to actually understand for people reading the code why it is needed in the specific place it is used.

I'll add a comment why sync roundtrip is required here. In general it dispatches all pending events to a server and reads a reply, it returns if it'll block.

Reply:
> All objects which closures wasn't assigned with quick_assign or some other things will go there, this is not part of sctk, but is a part of wayland-client, hence reexport.

So wayland-client panics internally when no closure is defined? That seems strange.

Reply:
No, it doesn't. Those events will go into fallback closure on dispatching event loop, like `         event_queue.sync_roundtrip(&mut (), |_, _, _| unreachable!())?;`. So they'll go into that `unreachable!()` thing.